File: DNA.VisualBasic/DNA.VisualBasic/OData/ODataEnableQueryAnalyzer.vb
Imports System.Collections.Immutable
Imports System.Threading
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class ODataEnableQueryAnalyzer
    Inherits DiagnosticAnalyzer

    Public Const DiagnosticId = "DNA101"
    Friend Shared ReadOnly Title As String = "OData warning"
    Friend Shared ReadOnly MessageFormat As String = "'{0}' method should be decorated with EnableQuery"
    Friend Const Category = "Syntax"

    Friend Shared Rule As New DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, True)

    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Rule)
        End Get
    End Property

    Public Overrides Sub Initialize(context As AnalysisContext)
        context.RegisterSyntaxNodeAction(AddressOf AnalyzeEnableQuery, SyntaxKind.FunctionStatement
                                         )
    End Sub

    Private Sub AnalyzeEnableQuery(context As SyntaxNodeAnalysisContext)
        If context.SemanticModel.Compilation.GetTypeByMetadataName("Microsoft.OData.Core.ODataAction") Is Nothing Then
            Return
        End If

        Dim root = context.Node
        Dim method As MethodStatementSyntax
        If TypeOf (root) Is MethodStatementSyntax Then
            method = TryCast(context.Node, MethodStatementSyntax)
        Else
            Return
        End If

        If method.Identifier.Value.ToString.ToLowerInvariant.StartsWith("[get") Or method.Identifier.Value.ToString.ToLowerInvariant.StartsWith("get") Then
            Dim descendantAttribs = method.AttributeLists.Where(Function(a) a.ToString.Contains("<EnableQuery>")).FirstOrDefault
            If descendantAttribs Is Nothing Then
                Dim diagn = Diagnostic.Create(Rule, root.GetLocation,
                      method.Identifier.Value.ToString)
                context.ReportDiagnostic(diagn)
            Else
                Return
            End If
        Else
            Return
        End If
    End Sub


End Class

File: DNA.CSharp/DNA.CSharp/OData/ODataEnableQueryAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace DNA.CSharp.OData
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class ODataEnableQueryAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "DNA101";
        internal static readonly string Title = "OData warning";
        internal static readonly string MessageFormat = "'{0}' method should be decorated with EnableQuery";
        internal const string Category = "Syntax";

        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(AnalyzeEnableQuery, SyntaxKind.MethodDeclaration);
        }


        private void AnalyzeEnableQuery(SyntaxNodeAnalysisContext context)
        {
            if (context.SemanticModel.Compilation.GetTypeByMetadataName("Microsoft.OData.Core.ODataAction") == null)
            {
                return;
            }

            var root = context.Node;
            MethodDeclarationSyntax method;
            if ((root) is MethodDeclarationSyntax)
            {
                method = context.Node as MethodDeclarationSyntax;
            }
            else
            {
                return;
            }

            if (method.Identifier.Value.ToString().ToLowerInvariant().StartsWith("get"))
            {
                var descendantAttribs = method.AttributeLists.Where(a => a.ToString().Contains("[EnableQuery]")).FirstOrDefault();
                if (descendantAttribs == null)
                {
                    var diagn = Diagnostic.Create(Rule, root.GetLocation(), method.Identifier.Value.ToString());
                    context.ReportDiagnostic(diagn);
                }
                else
                {
                    return;
                }
            }
            else
            {
                return;
            }
        }
    }
}